packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart

```diff
@@ -122,23 +117,95 @@ class TimelineController implements DisposableController {
   // this method fixes the regression without wasting resources to make the html
   // and flutter code 100% compatible.
   void htmlSelectTimelineEvent(TimelineEvent event) {
-    if (event == null || timeline.data.selectedEvent == event) return;
-    timeline.data.selectedEvent = event;
+    if (event == null || data.selectedEvent == event) return;
+    data.selectedEvent = event;
     _selectedTimelineEventNotifier.value = event;
   }
 
   Future<void> getCpuProfileForSelectedEvent() async {
-    final selectedEvent = timeline.data.selectedEvent;
+    final selectedEvent = data.selectedEvent;
     if (!selectedEvent.isUiEvent) return;
 
     await cpuProfilerController.pullAndProcessProfile(
       startMicros: selectedEvent.time.start.inMicroseconds,
       extentMicros: selectedEvent.time.duration.inMicroseconds,
     );
-    timeline.data.cpuProfileData = cpuProfilerController.dataNotifier.value;
+    data.cpuProfileData = cpuProfilerController.dataNotifier.value;
+  }
+
+  Future<double> get displayRefreshRate async {
+    final refreshRate =
+        await serviceManager.getDisplayRefreshRate() ?? defaultRefreshRate;
+    data?.displayRefreshRate = refreshRate;
+    return refreshRate;
   }
 
-  Future<void> loadOfflineData(OfflineData offlineData) async {
+  void selectFrame(TimelineFrame frame) {
+    if (frame == null || data.selectedFrame == frame || !hasStarted) {
+      return;
+    }
+    data.selectedFrame = frame;
+    _selectedFrameNotifier.value = frame;
+
+    data.selectedEvent = null;
+    _selectedTimelineEventNotifier.value = null;
+    data.cpuProfileData = null;
+    cpuProfilerController.reset();
+
+    if (debugTimeline && frame != null) {
+      final buf = StringBuffer();
+      buf.writeln('UI timeline event for frame ${frame.id}:');
+      frame.uiEventFlow.format(buf, '  ');
+      buf.writeln('\nUI trace for frame ${frame.id}');
+      frame.uiEventFlow.writeTraceToBuffer(buf);
+      buf.writeln('\nGPU timeline event frame ${frame.id}:');
+      frame.gpuEventFlow.format(buf, '  ');
+      buf.writeln('\nGPU trace for frame ${frame.id}');
+      frame.gpuEventFlow.writeTraceToBuffer(buf);
+      log(buf.toString());
+    }
+  }
+
+  void addFrame(TimelineFrame frame) {
+    data.frames.add(frame);
+  }
+
+  int vmStartRecordingMicros;
```

Comment:
should this be moved with the other fields?

Reply:
done